### PR TITLE
ENH add support for python 3 strings and proper null termination

### LINKED ISSUE #179: support unicode strings
I am making this issue to start a convo around unicode string support for python 3. TBH, this looks like a nightmare. 

Some notes:
- I checked the `cfitsio` docs and it does not appear that they have support for unicode. 
- Right now `fitsio` doesn't itself support the 'U' numpy type (which denotes unicode strings). 

So to make this happen, I think we have to put in manual conversions. On writing, this is straight forward since we would know it is a unicode type before hand and so could convert to a string of bytes (assuming we stick to a fixed length codec, which might cause overhead if we have to convert). 

Reading is the real nightmare though. Basically, if we see a column of bytes, IDK how we would know it was written from a column of unicode strings in python. I was thinking we could reserve `TSTRING` in the fits standard for `NPY_UNICODE` objects, but then if someone wants to write an array of python bytes data from `NPY_STRING`, we would have to use the `TBYTE` fits type. However, this would be confused with an array of unsigned one-byte integers, `NPY_UINT8`. 

Thus we have a many to one issue in that the FITS types are not expressive enough. 

This leaves us with truly awful solutions around monkey patching the header info to hold the extra information on which columns need unicode conversion on read. :donotwant:

So, unless, I am mistaken, we would have to basically write unicode into the FITS standard before we can support it in python 3. 

Thoughts @rainwoodman @esheldon @erykoff ?

### LINKED ISSUE #185: make fitsio strings match the standard
This issue has been brought up before.

cfitsio does not match the standard with regard to strings.  It converts null characters in the input strings to a space and always pads with spaces to the declared column size.  This is because cfitsio predates the standard and they want to maintain backward compatibility.

This padding with spaces is pervasive.  e.g. cfitsio will expand a table to have new rows if a single column in the table is extended, and the string columns are filled with spaces to the declared width.  So it is not a complete solution to just implement up our own string writing outside of cfitsio, because such operations as above will still result in spaces everywhere.

Probably the best thing to do is patch the sources.  an attempt was made in the nullstr branch that was not successful.

### PR DESCRIPTION
**Pull request by beckermr**
This PR contains a variety of updates for v1 including

- support for python 3 strings
- support in cfitsio for null terminated strings
- improved build system with setuptools and patches
- various bug fixes

Closes #179 
Closes #185

### CODE REVIEW

File: fitsio/fitsio_pywrap.c [hunk 1]
```diff
@@ -40,6 +40,13 @@ struct PyFITSObject {
     fitsfile* fits;
 };
 
+#ifdef FITSIO_PYWRAP_ALWAYS_NONSTANDARD_STRINGS
+static int fits_use_standard_strings(void)
+{
+    return 0;
+}
+#endif
+
```
Comment (beckermr) [inline fitsio/fitsio_pywrap.c]: @esheldon Why don't we want this macro around only the return value?

Comment (beckermr) [reply fitsio/fitsio_pywrap.c]: Ohhhhh. This function is here in case the version of fitsio doesn't have this symbol. NM.

Comment (esheldon) [reply fitsio/fitsio_pywrap.c]: I think that's correct

**Comment by beckermr** [comment]
OK. I round tripped all of the data from python 2, python 3, astropy, and both versions of fitsio. I got all of the things I expected. See the gist with the scripts here: https://gist.github.com/beckermr/897141e0a44ae2a192feebbfd42b40fa

**Comment by beckermr** [comment]
This is ready to merge on my end @esheldon.

**Pull request merged**